data-prepper-plugins/cloudwatch-logs/src/main/java/org/opensearch/dataprepper/plugins/sink/client/CloudWatchLogsService.java

```diff
@@ -0,0 +1,221 @@
+                                 final PluginMetrics pluginMetrics, final ThresholdCheck thresholdCheck, final int retryCount, final long backOffTimeBase) {
+
+        this.cloudWatchLogsClient = cloudWatchLogsClient;
+        this.buffer = buffer;
+        this.logGroup = cloudWatchLogsSinkConfig.getLogGroup();
+        this.logStream = cloudWatchLogsSinkConfig.getLogStream();
+        this.thresholdCheck = thresholdCheck;
+
+        this.retryCount = retryCount;
+        this.backOffTimeBase = backOffTimeBase;
+
+        this.bufferedEventHandles = new ArrayList<>();
+        this.logEventSuccessCounter = pluginMetrics.counter(NUMBER_OF_RECORDS_PUSHED_TO_CWL_SUCCESS);
+        this.requestFailCount = pluginMetrics.counter(REQUESTS_FAILED);
+        this.logEventFailCounter = pluginMetrics.counter(NUMBER_OF_RECORDS_PUSHED_TO_CWL_FAIL);
+        this.requestSuccessCount = pluginMetrics.counter(REQUESTS_SUCCEEDED);
+
+        reentrantLock = new ReentrantLock();
+
+        stopWatch = StopWatch.create();
+        stopWatchOn = false;
+    }
+
+    /**
+     * Function handles the packaging of events into log events before sending a bulk request to CloudWatchLogs.
+     * Implements simple conditional buffer. (Sends once batch size, request size in bytes, or time limit is reached)
+     * @param logs - Collection of Record events which hold log data.
+     */
+    public void output(final Collection<Record<Event>> logs) {
+        reentrantLock.lock();
+
+        try {
+            if (!stopWatchOn) {
+                startStopWatch();
+            }
+
+            for (Record<Event> singleLog: logs) {
+                String logJsonString = singleLog.getData().toJsonString();
+                int logLength = logJsonString.length();
+
+                if (thresholdCheck.isGreaterThanMaxEventSize(logLength + LOG_EVENT_OVERHEAD_SIZE)) {
+                    LOG.warn("Event blocked due to Max Size restriction! {Event Size: " + (logLength + LOG_EVENT_OVERHEAD_SIZE) + " bytes}");
+                    continue;
+                }
+
+                int bufferSizeWithOverHead = (buffer.getBufferSize() + (buffer.getEventCount() * LOG_EVENT_OVERHEAD_SIZE));
+                if ((thresholdCheck.isGreaterThanThresholdReached(getStopWatchTime(),  bufferSizeWithOverHead + logLength + LOG_EVENT_OVERHEAD_SIZE, buffer.getEventCount() + 1) && (buffer.getEventCount() > 0))) {
+                    pushLogs();
+                }
+
+                if (singleLog.getData().getEventHandle() != null) {
+                    bufferedEventHandles.add(singleLog.getData().getEventHandle());
+                }
+                buffer.writeEvent(logJsonString.getBytes());
+            }
+
+            runExitCheck();
+
+        } catch (InterruptedException e) {
+            LOG.error("Caught InterruptedException while attempting to publish logs!");
```

Comment:
I was not thinking of placing dlq logic there, but it could benefit as a "last resort", I will leave a "TODO" to address the dlq procedure for a future update of the plugin. Thank you for pointing it out, I also added a retry method so we re-attempt to process these logs as an InterruptedException does not correlate with our services timing out. This will be up for the next PR commit.